data/libs/Passengers.lua

```diff
@@ -0,0 +1,183 @@
+-- Copyright © 2008-2024 Pioneer Developers. See AUTHORS.txt for details
+-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+local Equipment = require 'Equipment'
+local EquipSet = require 'EquipSet'
+
+local utils = require 'utils'
+
+-- Module: Passengers
+--
+-- Helper utilities for working with passengers onboard a ship
+local Passengers = {}
+
+
+-- Function: CountOccupiedCabins
+--
+-- Return the number of occupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountOccupiedCabins(ship)
+	return ship["cabin_occupied_cap"] or 0
+end
+
+-- Function: CountFreeCabins
+--
+-- Return the number of unoccupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountFreeCabins(ship)
+	return (ship["cabin_cap"] or 0) - (ship["cabin_occupied_cap"] or 0)
+end
+
+-- Function: GetOccupiedCabins
+--
+-- Return a list of currently occupied cabin equipment items
+--
+---@param ship Ship
+---@return Equipment.CabinType[]
+function Passengers.GetOccupiedCabins(ship)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetNumPassengers() > 0
+	end)
+end
+
+-- Function: GetFreeCabins
+--
+-- Return a list of currently free passenger cabins
+--
+---@param ship Ship
+---@param numPassengers integer?
+---@return Equipment.CabinType[]
+function Passengers.GetFreeCabins(ship, numPassengers)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetFreeBerths() >= (numPassengers or 1)
+	end)
+end
+
+-- Function: CheckEmbarked
+--
+-- Validate how many of the given list of passengers are present on this ship
+--
+---@param ship Ship
+---@param passengers Character[]
+---@return integer
+function Passengers.CheckEmbarked(ship, passengers)
+	local occupiedCabins = Passengers.GetOccupiedCabins(ship)
+	local passengerLookup = utils.map_table(passengers, function(_, passenger)
+		return passenger, true
+	end)
+
+	local count = 0
+
+	for _, cabin in ipairs(occupiedCabins) do
+		for _, passenger in ipairs(cabin.passengers) do
+			if passengerLookup[passenger] then
+				count = count + 1
+			end
+		end
+	end
+
+	return count
+end
+
+-- Function: EmbarkPassenger
+--
+-- Load the given passenger onboard the ship, optionally to a specific cabin.
+--
+---@param ship Ship
+---@param passenger Character
+---@param cabin EquipType?
+function Passengers.EmbarkPassenger(ship, passenger, cabin)
+	---@cast cabin Equipment.CabinType?
+	if not cabin then
+		cabin = Passengers.GetFreeCabins(ship)[1]
+	end
+
+	if not cabin then
```

Comment:
Nitpick: sanity checks

What if the passed-in cabin is full?  Neither this function, nor the `cabin:AddPassenger()` method checks for overflow..